airbyte-integrations/bases/base-python/base_python/sdk/abstract_source.py

```diff
@@ -0,0 +1,148 @@
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+"""
+
+import copy
+from abc import ABC, abstractmethod
+from datetime import datetime
+from typing import Any, Iterator, Mapping, MutableMapping, Tuple
+
+from airbyte_protocol import (
+    AirbyteCatalog,
+    AirbyteConnectionStatus,
+    AirbyteMessage,
+    AirbyteRecordMessage,
+    AirbyteStateMessage,
+    AirbyteStream,
+    ConfiguredAirbyteCatalog,
+    ConfiguredAirbyteStream,
+    Status,
+    SyncMode,
+)
+from airbyte_protocol import Type as MessageType
+
+from base_python.integration import Source
+from base_python.logger import AirbyteLogger
+from base_python.sdk.streams.core import Stream, IncrementalStream
+
+
+class AbstractSource(Source, ABC):
+    def __init__(self):
+        super().__init__()
+
+    @abstractmethod
+    def check_connection(self, logger: AirbyteLogger, config: Mapping[str, Any]) -> Tuple[bool, Any]:
+        """
+        :return: A tuple of (boolean, error). If boolean is true, then we can connect to the underlying data source using the provided configuration.
+        Otherwise, the input config cannot be used to connect to the underlying data source, and the "error" object should describe what went wrong.
+        The error object will be cast to string to display the problem to the user.
+        """
+
+    @abstractmethod
+    def streams(self, config: Mapping[str, Any] = None) -> Mapping[str, Stream]:
+        """
+        :return: A mapping from stream name to the stream class representing that stream
+        """
+
+    @property
+    def name(self) -> str:
+        """Source name"""
+        return self.__class__.__name__
+
+    def discover(self, logger: AirbyteLogger, config: Mapping[str, Any]) -> AirbyteCatalog:
+        """Discover streams"""
+        streams = []
+
+        for name, stream in self.streams(config=config).items():
+            args = {
+                'name': name,
+                'json_schema': stream.get_json_schema(),
+            }
+
+            supported_sync_modes = [SyncMode.full_refresh]
+            if isinstance(stream, IncrementalStream):
+                supported_sync_modes.append(SyncMode.incremental)
+                args['source_defined_cursor'] = stream.source_defined_cursor
+                args['default_cursor_field'] = [stream.cursor_field] if isinstance(stream.cursor_field, str) else stream.cursor_field
+
+            args['supported_sync_modes'] = supported_sync_modes
+
+            streams.append(AirbyteStream(**args))
+
+        return AirbyteCatalog(streams=streams)
+
+    def check(self, logger: AirbyteLogger, config: Mapping[str, Any]) -> AirbyteConnectionStatus:
```

Comment:
just an idea i had while reading this... should each stream expose its own check method? it seems like a common failure case we have is that while the credentials are valid for some resources in an API they might not be valid for all of them. so in known cases for this, we could define custom check methods per stream.

so call check on the source and then each stream has the ability to expose its own check command which by default is a no op, but if it is present we execute it. then we just iterate over each stream and call check.

Reply:
yep, that's a nice one :)

Reply:
Good idea. Dovetails really nicely with #2014 and https://github.com/airbytehq/airbyte/issues/2364. I've created a follow up issue to do this: https://github.com/airbytehq/airbyte/issues/2765